### PR TITLE
No Cache

### PR DESCRIPTION
**Pull request by chriselrod**
See commit message for more details.

Setting the precompilation preferences to false, I get this on the latest release:
```julia
julia> using OrdinaryDiffEq

julia> function lorenz(du, u, p, t)
         du[1] = 10.0(u[2] - u[1])
         du[2] = u[1] * (28.0 - u[3]) - u[2]
         du[3] = u[1] * u[2] - (8 / 3) * u[3]
       end
lorenz (generic function with 1 method)

julia> function lorenz_oop(u, p, t)
         [10.0(u[2] - u[1]), u[1] * (28.0 - u[3]) - u[2], u[1] * u[2] - (8 / 3) * u[3]]
       end
lorenz_oop (generic function with 1 method)

julia> solver = Vern9();

julia> prob = ODEProblem(lorenz, [1.0; 0.0; 0.0], (0.0, 1.0));

julia> @time @eval solve(prob, solver);
  5.120141 seconds (5.44 M allocations: 345.486 MiB, 6.31% gc time, 100.00% compilation time)

julia> @benchmark solve($prob, $solver)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  12.554 μs …  2.393 ms  ┊ GC (min … max): 0.00% … 95.20%
 Time  (median):     13.646 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   15.343 μs ± 59.249 μs  ┊ GC (mean ± σ):  9.86% ±  2.54%

            ▂▃▄▆▆▇███▇▇▅▄▂▁                                    
  ▁▁▁▁▂▃▄▅▇████████████████▇▆▅▅▄▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  12.6 μs         Histogram: frequency by time        16.1 μs <

 Memory estimate: 28.34 KiB, allocs estimate: 259.
```
Vs this on this branch:
```julia
julia> using OrdinaryDiffEq

julia> function lorenz(du, u, p, t)
         du[1] = 10.0(u[2] - u[1])
         du[2] = u[1] * (28.0 - u[3]) - u[2]
         du[3] = u[1] * u[2] - (8 / 3) * u[3]
       end
lorenz (generic function with 1 method)

julia> function lorenz_oop(u, p, t)
         [10.0(u[2] - u[1]), u[1] * (28.0 - u[3]) - u[2], u[1] * u[2] - (8 / 3) * u[3]]
       end
lorenz_oop (generic function with 1 method)

julia> solver = Vern9();

julia> prob = ODEProblem(lorenz, [1.0; 0.0; 0.0], (0.0, 1.0));

julia> @time @eval solve(prob, solver);
  2.909396 seconds (6.35 M allocations: 407.016 MiB, 12.35% gc time, 99.99% compilation time)

julia> @benchmark solve($prob, $solver)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  10.631 μs …  2.528 ms  ┊ GC (min … max): 0.00% … 98.38%
 Time  (median):     11.649 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   12.996 μs ± 54.907 μs  ┊ GC (mean ± σ):  9.32% ±  2.19%

         ▂▁▂▃▅▅█▇▇█▇▆▄▂                                        
  ▁▁▂▃▅▆█████████████████▆▅▄▄▃▃▃▃▃▃▃▂▃▂▂▂▂▂▂▂▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  10.6 μs         Histogram: frequency by time        14.4 μs <

 Memory estimate: 22.09 KiB, allocs estimate: 259.
```
Note the compile time improvement (which will more generally translate to shorter precompile times), plus the runtime improvement owing to fewer unnecessary copies.
We have the same number of allocations, but those allocations are smaller due to the decreased cachesize.

This PR implements the optimization for both Vern7 and Vern9.

### CODE REVIEW

File: src/caches/verner_caches.jl [hunk 1]
```diff
@@ -243,18 +241,18 @@ function alg_cache(alg::Vern9, u, rate_prototype, ::Type{uEltypeNoUnits},
...
 end
 
 function alg_cache(alg::Vern9, u, rate_prototype, ::Type{uEltypeNoUnits},
                    ::Type{uBottomEltypeNoUnits}, ::Type{tTypeNoUnits}, uprev, uprev2, f, t,
                    dt, reltol, p, calck,
                    ::Val{false}) where {uEltypeNoUnits, uBottomEltypeNoUnits, tTypeNoUnits}
-    tab = Vern9Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
-    Vern9ConstantCache(tab)
+    # tab = Vern9Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
+    Vern9ConstantCache()
```
Comment (ChrisRackauckas) [inline src/caches/verner_caches.jl:257]:
```suggestion
    Vern9ConstantCache()
```

File: src/caches/verner_caches.jl [hunk 2]
```diff
@@ -243,18 +241,18 @@ function alg_cache(alg::Vern9, u, rate_prototype, ::Type{uEltypeNoUnits},
...
-               k16, utilde, tmp, rtmp, atmp, tab, alg.stage_limiter!, alg.step_limiter!,
+               k16, utilde, tmp, rtmp, atmp, alg.stage_limiter!, alg.step_limiter!,
                alg.thread)
 end
 
-struct Vern9ConstantCache{TabType} <: OrdinaryDiffEqConstantCache
-    tab::TabType
+struct Vern9ConstantCache <: OrdinaryDiffEqConstantCache
+    # tab::TabType
 end
```
Comment (ChrisRackauckas) [inline src/caches/verner_caches.jl:250]:
```suggestion
struct Vern9ConstantCache <: OrdinaryDiffEqConstantCache end
```

File: src/caches/verner_caches.jl [hunk 3]
```diff
@@ -220,7 +218,7 @@ function alg_cache(alg::Vern9, u, rate_prototype, ::Type{uEltypeNoUnits},
                    ::Type{uBottomEltypeNoUnits}, ::Type{tTypeNoUnits}, uprev, uprev2, f, t,
                    dt, reltol, p, calck,
                    ::Val{true}) where {uEltypeNoUnits, uBottomEltypeNoUnits, tTypeNoUnits}
-    tab = Vern9Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
+    # tab = Vern9Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
     k1 = zero(rate_prototype)
```
Comment (ChrisRackauckas) [inline src/caches/verner_caches.jl:222]:
```suggestion
    k1 = zero(rate_prototype)
```

File: src/caches/verner_caches.jl [hunk 4]
```diff
@@ -103,19 +103,18 @@ function alg_cache(alg::Vern7, u, rate_prototype, ::Type{uEltypeNoUnits},
...
 end
 
 function alg_cache(alg::Vern7, u, rate_prototype, ::Type{uEltypeNoUnits},
                    ::Type{uBottomEltypeNoUnits}, ::Type{tTypeNoUnits}, uprev, uprev2, f, t,
                    dt, reltol, p, calck,
                    ::Val{false}) where {uEltypeNoUnits, uBottomEltypeNoUnits, tTypeNoUnits}
-    tab = Vern7Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
-    Vern7ConstantCache(tab)
+    # tab = Vern7Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
+    Vern7ConstantCache()
```
Comment (ChrisRackauckas) [inline src/caches/verner_caches.jl:117]:
```suggestion
    Vern7ConstantCache()
```

File: src/caches/verner_caches.jl [hunk 5]
```diff
@@ -103,19 +103,18 @@ function alg_cache(alg::Vern7, u, rate_prototype, ::Type{uEltypeNoUnits},
...
     rtmp = uEltypeNoUnits === eltype(u) ? utilde : zero(rate_prototype)
     Vern7Cache(u, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, utilde, tmp, rtmp, atmp,
-               tab, alg.stage_limiter!, alg.step_limiter!, alg.thread)
+               alg.stage_limiter!, alg.step_limiter!, alg.thread)
 end
 
-struct Vern7ConstantCache{TabType} <: OrdinaryDiffEqConstantCache
-    tab::TabType
+struct Vern7ConstantCache <: OrdinaryDiffEqConstantCache
 end
```
Comment (ChrisRackauckas) [inline src/caches/verner_caches.jl:110]:
```suggestion
struct Vern7ConstantCache <: OrdinaryDiffEqConstantCache end
```

File: src/caches/verner_caches.jl [hunk 6]
```diff
@@ -86,7 +86,7 @@ function alg_cache(alg::Vern7, u, rate_prototype, ::Type{uEltypeNoUnits},
                    ::Type{uBottomEltypeNoUnits}, ::Type{tTypeNoUnits}, uprev, uprev2, f, t,
                    dt, reltol, p, calck,
                    ::Val{true}) where {uEltypeNoUnits, uBottomEltypeNoUnits, tTypeNoUnits}
-    tab = Vern7Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
+    # tab = Vern7Tableau(constvalue(uBottomEltypeNoUnits), constvalue(tTypeNoUnits))
     k1 = zero(rate_prototype)
```
Comment (ChrisRackauckas) [inline src/caches/verner_caches.jl:90]:
```suggestion
    k1 = zero(rate_prototype)
```

File: src/caches/verner_caches.jl [hunk 7]
```diff
@@ -76,7 +76,7 @@ end
     tmp::uType
     rtmp::rateType
     atmp::uNoUnitsType
-    tab::TabType
+    # tab::TabType
     stage_limiter!::StageLimiter
```
Comment (ChrisRackauckas) [inline src/caches/verner_caches.jl:80]:
```suggestion
    stage_limiter!::StageLimiter
```

File: Project.toml [hunk 8]
```diff
@@ -48,7 +48,7 @@ ArrayInterfaceCore = "0.1.22"
 ArrayInterfaceGPUArrays = "0.1, 0.2"
 ArrayInterfaceStaticArrays = "0.1"
 ArrayInterfaceStaticArraysCore = "0.1.2"
-DataStructures = "0.18"
+DataStructures = "0.18, 0.19"
```
Comment (ChrisRackauckas) [inline Project.toml:51]: was this tested? Odd this hasn't come up in a CompatHelper?

File: src/integrators/integrator_utils.jl [hunk 9]
```diff
@@ -61,10 +61,11 @@ function _savevalues!(integrator, force_save, reduce_size)::Tuple{Bool, Bool}
     saved, savedexactly = false, false
     !integrator.opts.save_on && return saved, savedexactly
     tdir_t = integrator.tdir * integrator.t
-    while !isempty(integrator.opts.saveat) && first(integrator.opts.saveat) <= tdir_t # Perform saveat
+    saveat = integrator.opts.saveat
+    while !isempty(saveat) && first(saveat) <= tdir_t # Perform saveat
         integrator.saveiter += 1
         saved = true
-        curt = integrator.tdir * pop!(integrator.opts.saveat)
+        curt = integrator.tdir * pop!(saveat)
```
Comment (ChrisRackauckas) [inline src/integrators/integrator_utils.jl:68]: I assume this doesn't make a major difference?

File: src/tableaus/verner_tableaus.jl [hunk 10]
```diff
@@ -1038,8 +1063,8 @@ struct Vern7Tableau{T, T2}
     btilde8::T
     btilde9::T
     btilde10::T
-    extra::Vern7ExtraStages{T, T2}
-    interp::Vern7InterpolationCoefficients{T}
+    # extra::Vern7ExtraStages{T, T2}
+    # interp::Vern7InterpolationCoefficients{T}
 end
```
Comment (ChrisRackauckas) [inline src/tableaus/verner_tableaus.jl:1068]:
```suggestion
end
```

File: src/tableaus/verner_tableaus.jl [hunk 11]
```diff
@@ -1108,17 +1133,17 @@ function Vern7Tableau(T::Type{<:CompiledFloats}, T2::Type{<:CompiledFloats})
     btilde9 = convert(T, 0.08131747232495111)
     btilde10 = convert(T, -0.02029518466335628)
 
-    extra = Vern7ExtraStages(T, T2)
-    interp = Vern7InterpolationCoefficients(T)
+    # extra = Vern7ExtraStages(T, T2)
+    # interp = Vern7InterpolationCoefficients(T)
```
Comment (ChrisRackauckas) [inline src/tableaus/verner_tableaus.jl:1138]:
```suggestion
```

File: src/tableaus/verner_tableaus.jl [hunk 12]
```diff
@@ -1227,14 +1252,14 @@ function Vern7Tableau(T, T2)
     btilde9 = convert(T, 181081444637946577 // 2226845467039736466)
     btilde10 = convert(T, -2152106665253777 // 106040260335225546)
 
-    extra = Vern7ExtraStages(T, T2)
-    interp = Vern7InterpolationCoefficients(T)
+    # extra = Vern7ExtraStages(T, T2)
+    # interp = Vern7InterpolationCoefficients(T)
```
Comment (ChrisRackauckas) [inline src/tableaus/verner_tableaus.jl:1257]:
```suggestion
```

File: src/tableaus/verner_tableaus.jl [hunk 13]
```diff
@@ -3376,8 +3401,8 @@ struct Vern9Tableau{T, T2}
     btilde14::T
     btilde15::T
     btilde16::T
-    extra::Vern9ExtraStages{T, T2}
-    interp::Vern9InterpolationCoefficients{T}
+    # extra::Vern9ExtraStages{T, T2}
+    # interp::Vern9InterpolationCoefficients{T}
 end
```
Comment (ChrisRackauckas) [inline src/tableaus/verner_tableaus.jl:3406]:
```suggestion
end
```

File: src/tableaus/verner_tableaus.jl [hunk 14]
```diff
@@ -3498,8 +3523,8 @@ function Vern9Tableau(T::Type{<:CompiledFloats}, T2::Type{<:CompiledFloats})
     btilde15 = convert(T, 0.030570139830827976)
     btilde16 = convert(T, -0.04834231373823958)
 
-    extra = Vern9ExtraStages(T, T2)
-    interp = Vern9InterpolationCoefficients(T)
+    # extra = Vern9ExtraStages(T, T2)
+    # interp = Vern9InterpolationCoefficients(T)
```
Comment (ChrisRackauckas) [inline src/tableaus/verner_tableaus.jl:3528]:
```suggestion
```

File: src/tableaus/verner_tableaus.jl [hunk 15]
```diff
@@ -3885,8 +3910,8 @@ function Vern9Tableau(T, T2)
     btilde16 = convert(T,
                        -1839190071060649887127895100784 // 38045139523510634351420875415397)
 
-    extra = Vern9ExtraStages(T, T2)
-    interp = Vern9InterpolationCoefficients(T)
+    # extra = Vern9ExtraStages(T, T2)
+    # interp = Vern9InterpolationCoefficients(T)
```
Comment (ChrisRackauckas) [inline src/tableaus/verner_tableaus.jl:3915]:
```suggestion
```

File: src/integrators/integrator_utils.jl (hunk 9 above)
Comment (chriselrod) [reply src/integrators/integrator_utils.jl:68]: Yeah, I just did that while looking at where all the copies came from.
I thought, maybe it's thinking that stuff possibly gets mutated on each iter and is thus reloading on each call.
But I didn't notice any difference.

File: Project.toml (hunk 8 above)
Comment (chriselrod) [reply Project.toml:51]: I was trying to dev DataStructures. This should be reverted. I eventually just changed the deved DataStructures version back to 0.18 instead.

Comment (ChrisRackauckas) [inline Project.toml:51]:
```suggestion
DataStructures = "0.18"
```

File: src/tableaus/low_order_rk_tableaus.jl [hunk 16]
```diff
@@ -677,12 +682,14 @@ function Tsit5ConstantCache(T::Type, T2::Type)
...
-                       r14, r22, r23, r24, r32, r33, r34, r42, r43, r44, r52, r53, r54, r62,
-                       r63, r64, r72, r73, r74)
+    # r11, r12, r13, r14, r22, r23, r24, r32, r33, r34, r42, r43, r44, r52, r53, r54, r62, r63, r64, r72, r73, r74 = Tsit5Interp(T)
+    Tsit5ConstantCacheActual(c1, c2, c3, c4, c5, c6, a21, a31, a32, a41, a42, a43, a51, a52,
+                             a53,
+                             a54, a61, a62, a63, a64, a65, a71, a72, a73, a74, a75, a76,
+                             btilde1,
+                             btilde2, btilde3, btilde4, btilde5, btilde6, btilde7)
+    # r11, r12, r13, r14, r22, r23, r24, r32, r33, r34, r42, r43, r44, r52, r53, r54, r62,
+    # r63, r64, r72, r73, r74)
```
Comment (ChrisRackauckas) [inline src/tableaus/low_order_rk_tableaus.jl]:
```suggestion
    Tsit5ConstantCacheActual(c1, c2, c3, c4, c5, c6, a21, a31, a32, a41, a42, a43, a51, a52,
                             a53,
                             a54, a61, a62, a63, a64, a65, a71, a72, a73, a74, a75, a76,
                             btilde1,
                             btilde2, btilde3, btilde4, btilde5, btilde6, btilde7)
```

**Comment by chriselrod** [comment]
The benefit for Tsi5 is much smaller; master:
```julia
julia> using OrdinaryDiffEq

julia> function lorenz(du, u, p, t)
         du[1] = 10.0(u[2] - u[1])
         du[2] = u[1] * (28.0 - u[3]) - u[2]
         du[3] = u[1] * u[2] - (8 / 3) * u[3]
       end
lorenz (generic function with 1 method)

julia> solver = Tsit5();

julia> prob = ODEProblem(lorenz, [1.0; 0.0; 0.0], (0.0, 1.0));

julia> @time @eval solve(prob, solver);
  1.802172 seconds (3.12 M allocations: 201.118 MiB, 13.85% gc time, 99.99% compilation time)

julia> @benchmark solve($prob, $solver)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  10.717 μs …  1.706 ms  ┊ GC (min … max): 0.00% … 96.58%
 Time  (median):     12.252 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   13.284 μs ± 39.159 μs  ┊ GC (mean ± σ):  7.07% ±  2.38%

                 ▁▃▄▇███▆▃▃                                    
  ▁▁▂▂▂▃▄▅▆▆▇▇█▇█████████████▇▆▅▄▄▄▃▃▃▃▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  10.7 μs         Histogram: frequency by time        15.2 μs <

 Memory estimate: 25.50 KiB, allocs estimate: 293.
```
PR:
```julia
julia> using OrdinaryDiffEq

julia> function lorenz(du, u, p, t)
         du[1] = 10.0(u[2] - u[1])
         du[2] = u[1] * (28.0 - u[3]) - u[2]
         du[3] = u[1] * u[2] - (8 / 3) * u[3]
       end
lorenz (generic function with 1 method)

julia> solver = Tsit5();

julia> prob = ODEProblem(lorenz, [1.0; 0.0; 0.0], (0.0, 1.0));

julia> @time @eval solve(prob, solver);
  1.725814 seconds (3.34 M allocations: 215.145 MiB, 14.87% gc time, 99.99% compilation time)

julia> @benchmark solve($prob, $solver)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  10.659 μs …  1.851 ms  ┊ GC (min … max): 0.00% … 97.76%
 Time  (median):     12.141 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   13.294 μs ± 42.939 μs  ┊ GC (mean ± σ):  7.75% ±  2.38%

                   ▂▅▆█▇▅▃▂▁                                   
  ▁▁▁▂▃▃▄▅▆▆▆▆▆▅▅▅███████████▆▅▅▄▄▄▄▄▃▃▃▃▃▂▃▃▂▂▂▂▂▁▂▁▁▁▁▁▁▁▁▁ ▃
  10.7 μs         Histogram: frequency by time        14.8 μs <

 Memory estimate: 24.62 KiB, allocs estimate: 293.
```

**Comment by chriselrod** [comment]
PR:
```julia
julia> using StaticArrays, OrdinaryDiffEq
WARNING: using StaticArrays.pop in module Main conflicts with an existing identifier.

julia> function lorenz_oop(u, p, t)
         SA[10.0(u[2] - u[1]), u[1] * (28.0 - u[3]) - u[2], u[1] * u[2] - (8 / 3) * u[3]]
       end
lorenz_oop (generic function with 1 method)

julia> prob = ODEProblem(lorenz_oop, SA[1.0; 0.0; 0.0], (0.0, 1.0));

julia> solver = Tsit5();

julia> @time @eval solve(prob, solver);
  1.227860 seconds (2.53 M allocations: 164.619 MiB, 19.70% gc time, 99.60% compilation time)

julia> @benchmark solve($prob, $solver)
BenchmarkTools.Trial: 10000 samples with 7 evaluations.
 Range (min … max):  4.592 μs … 158.457 μs  ┊ GC (min … max): 0.00% … 93.49%
 Time  (median):     4.721 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   4.981 μs ±   5.217 μs  ┊ GC (mean ± σ):  3.90% ±  3.61%

  ▂▅▆▇███▇▇▆▅▄▃▂▂▂▁▂▁▂▂▁▂▂▂▂▂▁▁▂ ▁▁▁▁▁                        ▃
  ██████████████████████████████████████▇▇▇▅▅▅▃▃▄▅▄▁▄▁▅▃▄▁▅▅▄ █
  4.59 μs      Histogram: log(frequency) by time      5.81 μs <

 Memory estimate: 9.73 KiB, allocs estimate: 48.
```
latest release:
```julia
julia> using StaticArrays, OrdinaryDiffEq
WARNING: using StaticArrays.pop in module Main conflicts with an existing identifier.

julia> function lorenz_oop(u, p, t)
         SA[10.0(u[2] - u[1]), u[1] * (28.0 - u[3]) - u[2], u[1] * u[2] - (8 / 3) * u[3]]
       end
lorenz_oop (generic function with 1 method)

julia> prob = ODEProblem(lorenz_oop, SA[1.0; 0.0; 0.0], (0.0, 1.0));

julia> solver = Tsit5();

julia> @time @eval solve(prob, solver);
  1.255036 seconds (2.43 M allocations: 158.278 MiB, 19.01% gc time, 99.60% compilation time)

julia> @benchmark solve($prob, $solver)
BenchmarkTools.Trial: 10000 samples with 7 evaluations.
 Range (min … max):  4.734 μs … 141.349 μs  ┊ GC (min … max): 0.00% … 92.70%
 Time  (median):     4.897 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   5.162 μs ±   4.834 μs  ┊ GC (mean ± σ):  3.69% ±  3.80%

   ▂▄▆▇███▇▇▆▅▄▃▂▁▁▁▂▁▁▁▁▁▂▁▁▁▁▁▂▂▁▂▁▁▁▁                      ▃
  ▇███████████████████████████████████████▇▇█▇▆▄▆▅▄▅▅▅▅▁▄▄▃▅▆ █
  4.73 μs      Histogram: log(frequency) by time      6.03 μs <

 Memory estimate: 10.70 KiB, allocs estimate: 48.
```

**Comment by chriselrod** [comment]
Vern9 sees a big benefit for both ip and oop; latest release:
```julia
julia> using StaticArrays, OrdinaryDiffEq
WARNING: using StaticArrays.pop in module Main conflicts with an existing identifier.

julia> function lorenz_oop(u, p, t)
         SA[10.0(u[2] - u[1]), u[1] * (28.0 - u[3]) - u[2], u[1] * u[2] - (8 / 3) * u[3]]
       end
lorenz_oop (generic function with 1 method)

julia> prob = ODEProblem(lorenz_oop, SA[1.0; 0.0; 0.0], (0.0, 1.0));

julia> solver = Vern9();

julia> @time @eval solve(prob, solver);
  2.181060 seconds (3.48 M allocations: 217.965 MiB, 11.79% gc time, 99.77% compilation time)

julia> @benchmark solve($prob, $solver)
BenchmarkTools.Trial: 10000 samples with 6 evaluations.
 Range (min … max):  5.484 μs … 208.429 μs  ┊ GC (min … max): 0.00% … 93.22%
 Time  (median):     5.893 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.317 μs ±   7.779 μs  ┊ GC (mean ± σ):  5.10% ±  4.03%

       ▁▅▆▇▇██▆▃                                               
  ▁▁▁▃▆█████████▇▅▃▃▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃^G
  5.48 μs         Histogram: frequency by time         7.8 μs <

 Memory estimate: 15.22 KiB, allocs estimate: 38.
```
PR:
```julia
julia> using StaticArrays, OrdinaryDiffEq
WARNING: using StaticArrays.pop in module Main conflicts with an existing identifier.

julia> function lorenz_oop(u, p, t)
         SA[10.0(u[2] - u[1]), u[1] * (28.0 - u[3]) - u[2], u[1] * u[2] - (8 / 3) * u[3]]
       end
lorenz_oop (generic function with 1 method)

julia> prob = ODEProblem(lorenz_oop, SA[1.0; 0.0; 0.0], (0.0, 1.0));

julia> solver = Vern9();

julia> @time @eval solve(prob, solver);
  1.870612 seconds (4.03 M allocations: 255.161 MiB, 15.23% gc time, 99.74% compilation time)

julia> @benchmark solve($prob, $solver)
BenchmarkTools.Trial: 10000 samples with 8 evaluations.
 Range (min … max):  3.833 μs … 135.757 μs  ┊ GC (min … max): 0.00% … 93.96%
 Time  (median):     3.990 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   4.199 μs ±   4.522 μs  ┊ GC (mean ± σ):  3.87% ±  3.49%

      ▁▅▆██▇▇▅▅▄▄▃▁                                            
  ▁▂▄▅██████████████▆▅▄▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂▂▂▁▁▁▂▁▁▁▁▁▁▁▁▁▁▁ ▃
  3.83 μs         Histogram: frequency by time        4.63 μs <

 Memory estimate: 9.00 KiB, allocs estimate: 38.
```

**Comment by chriselrod** [comment]
I should actually time precompilation, but it is notably faster on this PR.
Latest release
```julia
julia> info_cachefile("OrdinaryDiffEq")
Contents of /home/chriselrod/.julia/compiled/v1.10/OrdinaryDiffEq/DlSvy_lNMJ5.so:
  modules: Any[OrdinaryDiffEq]
  2871 external methods
  17484 new specializations of external methods (Base 52.9%, UnPack 15.2%, Base.Broadcast 14.2%, ...)
  844 external methods with new roots
  28776 external targets
  29269 edges
  file size:   143549408 (136.899 MiB)
  Segment sizes (bytes):
  system:      27307504 ( 23.69%)
  isbits:      84767781 ( 73.52%)
  symbols:       136012 (  0.12%)
  tags:          383679 (  0.33%)
  relocations:  2619334 (  2.27%)
  gvars:          22576 (  0.02%)
  fptrs:          56576 (  0.05%)
```
PR:
```julia
julia> info_cachefile("OrdinaryDiffEq")
Contents of /home/chriselrod/.julia/compiled/v1.10/OrdinaryDiffEq/DlSvy_hyQim.so:
  modules: Any[OrdinaryDiffEq]
  2871 external methods
  16375 new specializations of external methods (Base 53.6%, Base.Broadcast 15.2%, UnPack 12.3%, ...)
  844 external methods with new roots
  27667 external targets
  28637 edges
  file size:   140992328 (134.461 MiB)
  Segment sizes (bytes):
  system:      26685688 ( 23.48%)
  isbits:      83864397 ( 73.78%)
  symbols:       135365 (  0.12%)
  tags:          373715 (  0.33%)
  relocations:  2536109 (  2.23%)
  gvars:          22472 (  0.02%)
  fptrs:          56576 (  0.05%)
```
Note `UnPack`'s decrease, and the smaller image size.

**Comment by chriselrod** [comment]
Naively, this failure looks real? https://github.com/SciML/OrdinaryDiffEq.jl/actions/runs/4031769906/jobs/6931258704

File: src/dense/interpolants.jl [hunk 17]
```diff
@@ -366,11 +366,10 @@ simplifying assumption
 
 Ch. Tsitouras
 """
-@def tsit5unpack begin if typeof(cache) <: OrdinaryDiffEqMutableCache
-    @unpack r11, r12, r13, r14, r22, r23, r24, r32, r33, r34, r42, r43, r44, r52, r53, r54, r62, r63, r64, r72, r73, r74 = cache.tab
-else
-    @unpack r11, r12, r13, r14, r22, r23, r24, r32, r33, r34, r42, r43, r44, r52, r53, r54, r62, r63, r64, r72, r73, r74 = cache
-end end
+@def tsit5unpack begin
+    var"#T#" = recursive_unitless_bottom_eltype(y₁)
```
Comment (ChrisRackauckas) [inline src/dense/interpolants.jl:370]: I think the types are all missing the `constvalue` part which is crucial for dropping units, ADness, and complex-ness

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/OrdinaryDiffEq.jl/actions/runs/4031769902/jobs/6931259103#step:6:734 that looks like a real error

**Comment by chriselrod** [comment]
> https://github.com/SciML/OrdinaryDiffEq.jl/actions/runs/4031769902/jobs/6931259103#step:6:734 that looks like a real error

Yes, that's from abusing `@generated`.

**Comment by oscardssmith** [comment]
@ChrisRackauckas can you explain the theory behind this PR? I feel very out of the loop here. What is `Base.Experimental.silence!`, and why so many `@generated` functions?

**Comment by chriselrod** [comment]
Hmm, I guess these record history?
```julia
julia> Base.isbitstype(Tracker.TrackedReal{Float32})
false
```
If they're properly mutable, then we do actually need to allocate them fresh.

**Comment by chriselrod** [comment]
> can you explain the theory behind this PR?

@oscardssmith, LLVM spent all its time compiling the copying of constants like tableaus for Vern9.
So instead of caching the creation of structs holding these constants and copying them around, we just inline the constants to where they are used.

**Comment by ChrisRackauckas** [comment]
> What is Base.Experimental.silence!

It's Jeff's new experimental stack trace silencing feature. It should probably be split to a separate PR, but it's fine because it's not even in a Base PR yet.

**Comment by ChrisRackauckas** [comment]
> If they're properly mutable, then we do actually need to allocate them fresh.

None of the coefficients should be tracked values though, that should be handled by the `constvalue` function which should call `value`. The issue here is likely that https://github.com/SciML/SciMLSensitivity.jl/blob/master/src/tracker.jl#L15-L18 needs to be made into an extension in DiffEqBase instead.

**Comment by chriselrod** [comment]
> > If they're properly mutable, then we do actually need to allocate them fresh.
> 
> None of the coefficients should be tracked values though, that should be handled by the `constvalue` function which should call `value`. The issue here is likely that https://github.com/SciML/SciMLSensitivity.jl/blob/master/src/tracker.jl#L15-L18 needs to be made into an extension in DiffEqBase instead.

Okay, so these tableau functions should never be called with `TrackedReal`.

We should still probably switch to `Base.@assume_effects :foldable` in Julia 1.8, only using some more hacky thing for older Julia versions.

**Comment by ChrisRackauckas** [comment]
SciMLSensitivity.jl in general should be loaded any time an AD package is loaded, but the circular dependency issue makes that hard. That's what I posted here:

https://discourse.julialang.org/t/package-extensions-for-julia-1-9/93397/12?u=chrisrackauckas

Making the following:

https://github.com/SciML/SciMLSensitivity.jl/blob/master/src/tracker.jl
https://github.com/SciML/SciMLSensitivity.jl/blob/master/src/zygote.jl
https://github.com/SciML/SciMLSensitivity.jl/blob/master/src/reversediff.jl

into DiffEqBase extension packages would probably be a quick fix for now.

**Comment by chriselrod** [comment]
All the AD backends are fairly intermingled, e.g. `automatic_sensealg_choice` or `inplace_vjp`.

**Comment by ChrisRackauckas** [comment]
That's not what I meant 😅 . I meant loading the whole SciMLSensitivity if any reverse mode AD is loaded.

**Comment by chriselrod** [comment]
Oh, sorry, that makes way more sense.

I guess that requires a major version bump on DiffEqBase to avoid overwriting the methods?

If it weren't for this PR, what I'd have done is leave the code in SciMLSensitivity (while copying it into DiffEqBase), but only load it if `!isdefined(Base, :get_extension)`.

This way in Julia versions without `get_extension`, nothing changes.
In Julia >= 1.9, it'd be an extension package.

Of course, that doesn't fix this particular issue, as this issue is only in Julia <= 1.7, where `Base.@constprop` isn't defined.

**Comment by ChrisRackauckas** [comment]
> I guess that requires a major version bump on DiffEqBase to avoid overwriting the methods?

If they are the same methods it's not breaking, you just get a warning, so I think it could be fine?

**Comment by chriselrod** [comment]
Unfortunately, that isn't currently legal.
I.e., we can't have load `OrdinaryDiffEq/ext/SciMLSensitivityExt.jl` that does `using SciMLSensitivity` in response to one of `SciMLSensitivityExt`'s dependencies getting loaded.
Pkg doesn't know how to handle the cycle `ReverseDiff` -> load `SciMLSensitivityExt` -> load `SciMLSensitivity` -> load `ReverseDiff` -> load `SciMLSensitivityExt` -> errors on cycle detection.

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/SciMLSensitivity.jl/pull/780 is what needs to get finished to complete this. We can worry about automatically bringing in SciMLSensitivity later, but with this at least the correct errors are always thrown and the piracy is eliminated.

**Comment by ChrisRackauckas** [comment]
Okay, fingers crossed it may work this time.

**Comment by ChrisRackauckas** [comment]
Now fix your fork lol

**Comment by chriselrod** [comment]
> Now fix your fork lol

Done:
https://github.com/chriselrod/OrdinaryDiffEq.jl/pulls

**Pull request merged**